pkg/prometheus/agent/daemonset.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2023 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusagent
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/slices"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/prometheus-operator/prometheus-operator/pkg/k8sutil"
+	"github.com/prometheus-operator/prometheus-operator/pkg/operator"
+	prompkg "github.com/prometheus-operator/prometheus-operator/pkg/prometheus"
+)
+
+func makeDaemonSet(
+	name string,
+	p monitoringv1.PrometheusInterface,
+	config *prompkg.Config,
+	cg *prompkg.ConfigGenerator,
+	tlsSecrets *operator.ShardedSecret,
+) (*appsv1.DaemonSet, error) {
+	cpf := p.GetCommonPrometheusFields()
+	objMeta := p.GetObjectMeta()
+
+	if cpf.PortName == "" {
+		cpf.PortName = prompkg.DefaultPortName
+	}
+
+	// We need to re-set the common fields because cpf is only a copy of the original object.
+	// We set some defaults if some fields are not present, and we want those fields set in the original Prometheus object before building the DaemonSetSpec.
+	p.SetCommonPrometheusFields(cpf)
+
+	spec, err := makeDaemonSetSpec(p, config, cg, tlsSecrets)
+	if err != nil {
+		return nil, fmt.Errorf("make DaemonSet spec: %w", err)
+	}
+	daemonSet := &appsv1.DaemonSet{Spec: *spec}
+
+	operator.UpdateObject(
+		daemonSet,
+		operator.WithName(name),
+		operator.WithAnnotations(objMeta.GetAnnotations()),
+		operator.WithAnnotations(config.Annotations),
+		operator.WithLabels(objMeta.GetLabels()),
+		operator.WithLabels(map[string]string{
+			prompkg.PrometheusNameLabelName: objMeta.GetName(),
+			prompkg.PrometheusModeLabeLName: prometheusMode,
+		}),
+		operator.WithLabels(config.Labels),
+		operator.WithManagingOwner(p),
```

Comment:
;)
```suggestion
		operator.WithManagingOwner(p),
		operator.WithoutKubectlAnnotations(),
```

Reply:
Not directly related to this change, but do you think we need an unit test for annotation and label? Something like https://github.com/prometheus-operator/prometheus-operator/blob/680654c9d110ab90c214f3185ca85926f1c72f6f/pkg/prometheus/server/statefulset_test.go#L79 but doesn't have `prometheus-operator-input-hash` and `operator.prometheus.io/shard`.

I think this is needed because this is the part where daemonset is different from statefulset (both agent and server).

Reply:
Yes a unit test would be good.